File: tests/fixtures/open_redirect/ruby/safe_relative_redirect.rb
# Safe: query arg routed through `ensure_relative_url` which enforces a
# leading `/` and rejects scheme-prefixed values (relative-only path).
class HomeController
  def ensure_relative_url(raw)
    return '/' unless raw.is_a?(String)
    return '/' unless raw.start_with?('/')
    return '/' if raw.start_with?('//')
    raw
  end

  def jump
    target = params[:next]
    safe = ensure_relative_url(target)
    redirect_to safe
  end
end

File: tests/fixtures/open_redirect/php/safe_relative_redirect.php
<?php
// Safe: $_GET['next'] routed through `ensure_relative_url` which enforces
// a leading `/` and rejects scheme-prefixed values (relative-only path).
function ensure_relative_url($raw) {
    if (!is_string($raw) || strpos($raw, '/') !== 0 || strpos($raw, '//') === 0) {
        return '/';
    }
    return $raw;
}

$next = $_GET['next'];
$safe = ensure_relative_url($next);
header("Location: " . $safe);
